kernel/src/memory/paging/mod.rs

```diff
@@ -0,0 +1,290 @@
+        }
+    }
+}
+
+/// An entry in a page table
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(C)] // Just in case
+pub struct PageTableEntry(u64);
+
+impl PageTableEntry {
+    pub fn set_unused(&mut self) {
+        self.0 = 0;
+    }
+
+    pub fn flags(&self) -> EntryFlags {
+        EntryFlags::from_bits_truncate(self.0)
+    }
+
+    pub fn physical_address(&self) -> Option<PhysicalAddress> {
+        if self.flags().contains(self::EntryFlags::PRESENT) {
+            Some(PhysicalAddress(self.0 as usize & 0x000FFFFF_FFFFF000)) // Mask out the flag bits
+        } else {
+            None
+        }
+    }
+
+    pub fn set(&mut self, physical_address: PhysicalAddress, flags: EntryFlags) {
+        // Check that the physical address is page aligned
+        assert_eq!(
+            physical_address.0 & 0xFFF,
+            0,
+            "Physical address 0x{:x} not page aligned!",
+            physical_address.0,
+        );
+
+        // Check that physical address is correctly sign extended
+        let bit_47 = (physical_address.0 >> 48) & 1;
+        if bit_47 == 1 {
+            assert_eq!(
+                physical_address.0 >> 48,
+                0xFFFF,
+                "Physical address 0x{:x} is not correctly sign extended!",
+                physical_address.0,
+            )
+        } else {
+            assert_eq!(
+                physical_address.0 >> 48,
+                0,
+                "Physical address 0x{:x} is not correctly sign extended!",
+                physical_address.0,
+            )
+        }
+
+        self.0 = (physical_address.0 as u64) | flags.bits();
+    }
+}
+
+bitflags! {
+    pub struct EntryFlags: u64 {
+        /// Whether the page is present in memory
+        const PRESENT = 1;
+        /// Whether the page is writable or read only
+        const WRITABLE = 1 << 1;
+        /// Whether ring 3 processes can access this page -- in theory. As of meltdown, this bit is
+        /// essentially useless, except on possibly newer CPUs with fixes in place
+        const USER_ACCESSIBLE = 1 << 2;
+        /// If this bit is set, writes to this page go directly to memory
+        const WRITE_DIRECT = 1 << 3;
+        /// Do not use cache for this page
+        const NO_CACHE = 1 << 4;
+        /// Set by the CPU when this page has been accessed
+        const ACCESSED = 1 << 5;
+        /// Set by the CPU when this page is written to
+        const DIRTY = 1 << 6;
+        /// Whether this page is a huge page. 0 in P1 and P4, but sets this as a 1GiB page in P3
+        /// and a 2MiB page in P2
+        const HUGE_PAGE = 1 << 7;
+        /// If set, this page will not be flushed in the TLB. PGE bit in CR4 must be set.
+        const GLOBAL = 1 << 8;
+        /// Do not allow executing code from this page. NXE bit in EFER must be set.
+        const NO_EXECUTE = 1 << 63;
+    }
+}
+
+/// A trait that indicates a type represents a page table level
+pub trait TableLevel {}
+
+pub enum Level4 {}
```

Comment:
Do enums get their index optimized away when no values?

Reply:
Enums like this cannot ever exist as they have no values. This is never constructed, it is just a type level thing.